src/sys/windows/sourcerawsocket.rs

```diff
@@ -0,0 +1,128 @@
+use crate::{event, poll, Interests, Registry, Token};
+
+use std::io;
+use std::os::windows::io::{AsRawSocket, RawSocket};
+use std::sync::{Arc, Mutex};
+
+use super::selector::SockState;
+
+/// Adapter for [`RawSocket`] providing an [`event::Source`] implementation.
+///
+/// `SourceRawSocket` enables registering any type with an RawSocket with [`Poll`].
+///
+/// Note that `SourceRawSocket` takes a `&RawSocket`. This is because `SourceRawSocket` **does
+/// not** take ownership of the RawSocket. Specifically, it will not manage any
+/// lifecycle related operations, such as closing the RawSocket on drop. It is expected
+/// that the `SourceRawSocket` is constructed right before a call to
+/// [`Registry::register`]. See the examples for more detail.
+///
+/// [`event::Source`]: crate::event::Source
+/// [`Poll`]: crate::Poll
+/// [`Registry::register`]: crate::Registry::register
+///
+/// # Examples
+///
+/// Basic usage.
+///
+/// ```
+/// # use std::error::Error;
+/// # fn main() -> Result<(), Box<dyn Error>> {
+/// use mio::{Interests, Poll, Token};
+/// use mio::windows::SourceRawSocket;
+///
+/// use std::os::windows::io::AsRawSocket;
+/// use std::net::TcpListener;
+///
+/// // Bind a std listener
+/// let listener = TcpListener::bind("127.0.0.1:0")?;
+///
+/// let poll = Poll::new()?;
+/// let registry = poll.registry().clone();
+///
+/// // Register the listener
+/// registry.register(
+///     &SourceRawSocket::new(&listener.as_raw_socket()),
+///     Token(0),
+///     Interests::READABLE)?;
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// Implementing [`event::Source`] for a custom type backed by a [`RawSocket`].
+///
+/// ```
+/// use mio::{event, Interests, Registry, Token};
+/// use mio::windows::SourceRawSocket;
+///
+/// use std::os::windows::io::RawSocket;
+/// use std::io;
+///
+/// # #[allow(dead_code)]
+/// pub struct MyIo {
+///     raw_socket: RawSocket,
+/// }
+///
+/// impl event::Source for MyIo {
+///     fn register(&self, registry: &Registry, token: Token, interests: Interests)
+///         -> io::Result<()>
+///     {
+///         SourceRawSocket::new(&self.raw_socket).register(registry, token, interests)
```

Comment:
This usage is rather expensive if we're creating a new `Arc<Mutex<..>>` each call to register. If we can't do without the `SockState` we might need to go for a different design.

Reply:
The problem is we need to simulate ET behavior. Then we do need SockState to maintain what interests have been read

Reply:
That is fine, but would the example still work then? It throws about the same after each call to (re)register doesn't it?

Reply:
Indeed. You are right. I cannot hold all the actions user took on the RawSocket.

Reply:
Then the library will need to take level trigger approach rather than edge trigger

Reply:
or just remove SourceRawSocket

Reply:
Let's remove it for now then and we'll find another approach later.

Reply:
Removed